File: algorithms/src/algo/strings/KnuthMorrisPratt2.java
package algo.strings;

import algovue.annotations.Block;
import algovue.annotations.EolComment;
import algovue.annotations.Indexes;
import algovue.annotations.MethodComment;

public class KnuthMorrisPratt2 {

    @MethodComment("Preprocess the pattern (calculate lps[] array)")
    int[] computeLPSArray(String pat) {
        int M = pat.length();
        int[] lps = new int[M];
        lps[0] = 0; // lps[0] is always 0

        int _0;

        // length of the previous longest prefix suffix
        @Indexes({"pat", "lps"})
        int j = 0;
        @Indexes({"pat", "lps"}) int i = 1;

        int _1;

        // the loop calculates lps[i] for i = 1 to M-1
        while (i < M) {
            while (pat.charAt(i) == pat.charAt(j)) {
                @Block(text = "Match at position i", colors = {"#F0FFF0", "D0FFD0"}) int $a;
                @EolComment("Fill lps as long as there is a match. Index 0 means, that 1 char matched, thus +1") int _c2;
                lps[i] = j + 1;
                j++;
                i++;
                if (i == M) {
                    return lps;
                }
            }

            @Block(text = "Mismatch at position i", colors = {"FFFAFA", "FFF0F0"}) int $b;
            if (j == 0) {
                @Block(colors = {"#F4F0FF", "E0D0FF"}) int $b0;
                @EolComment("For mismatch, put match length by 1 less, than in the case of match") int _c2;
                lps[i] = j;
                i++;
            } else {
                @Block(colors = {"#F0F8FF", "A0D0FF"}) int $b1;
                @EolComment("Note, that i is NOT incremented; only pointer j in pattern is slashed back") int _b1;
                j = lps[j - 1];
            }
        }

        int _2;
        return lps;
    }


    @MethodComment("Knuth-Morris-Pratt algorithm")
    private int kmp(String txt, String pat) {

        int N = txt.length();
        int M = pat.length();
        if (M <= 0) return -1;
        int _emptyLine0;

        int[] lps = computeLPSArray(pat);
        @Indexes({"pat", "lps"}) int j = 0;
        @Indexes("txt") int i = 0;

        int _emptyLine1;
        while (i < N) {
            while (txt.charAt(i) == pat.charAt(j)) {
                @Block(text = "Match at position i", colors = {"#F0FFF0", "D0FFD0"}) int $a;
                j++;
                i++;
                if (j == M) {
                    return i - j;
                }
            }
            @Block(text = "Mismatch at position i", colors = {"FFFAFA", "FFF0F0"}) int $b;
            if (j == 0) {
                @Block(colors = {"#F4F0FF", "E0D0FF"}) int $b0;
                @EolComment("Mismatch at the very first pattern char; cannot shift back in pattern, so start a new match") int _c2;
                i++;
            } else {
                @Block(colors = {"#F0F8FF", "A0D0FF"}) int $b1;
                @EolComment("We know that j charaters match. Note, that i is NOT incremented.") int _c2;
                j = lps[j - 1];
            }
        }

        int _emptyLine2;
        return -1;
    }

    int index = kmp("ABABDABACDABABCABAB", "ABABCABAB");
}

File: v1/algo/strings/knuth_morris_pratt2.js
test = function() {

    const computeLPSArray = vm.functionDeclaration(
        'computeLPSArray',
        [vm.variable('pat')],
        vm.sequenceStatement([
            vm.assignment(
                vm.varWrite('M'),
                vm.functionCall('length', [], 'pat')
            ),
            vm.assignment(
                vm.varWrite('lps'),
                vm.newIntArray(vm.variable('M'))
            ),
            vm.assignment(
                vm.arrItemWrite('lps', vm.number(0)),
                vm.number(0)
            ),
            vm.standAloneComment(),
            vm.assignment(
                vm.varWrite('j', {"role":"index", "targetArrays":["pat", "lps"]}),
                vm.number(0)
            ),
            vm.assignment(
                vm.varWrite('i', {"role":"index", "targetArrays":["pat", "lps"]}),
                vm.number(1)
            ),
            vm.standAloneComment(),
            vm.whileStatement(
                vm.expression(vm.lt(), vm.variable('i'), vm.variable('M')),
                vm.sequenceStatement([
                    vm.whileStatement(
                        vm.expression(vm.eq(), vm.arrItem('pat', vm.variable('i')), vm.arrItem('pat', vm.variable('j'))),
                        vm.sequenceStatement([
                            vm.group('// Match at position i','#F0FFF0', 'D0FFD0',
                                [
                                    vm.assignment(
                                        vm.arrItemWrite('lps', vm.variable('i')),
                                        vm.expression(vm.plus(), vm.variable('j'), vm.number(1)), ' // Fill lps as long as there is a match. Index 0 means, that 1 char matched, thus +1'
                                    ),
                                    vm.assignment(
                                        undefined,
                                        vm.varPostOp('j', true)
                                    ),
                                    vm.assignment(
                                        undefined,
                                        vm.varPostOp('i', true)
                                    ),
                                    vm.ifStatement(
                                        vm.expression(vm.eq(), vm.variable('i'), vm.variable('M')),
                                        vm.sequenceStatement([
                                            vm.returnStatement(vm.variable('lps'))
                                        ]),
                                        undefined
                                    )
                                ])
                        ])
                    ),
                    vm.group('// Mismatch at position i','FFFAFA', 'FFF0F0',
                        [
                            vm.ifStatement(
                                vm.expression(vm.eq(), vm.variable('j'), vm.number(0)),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#F4F0FF', 'E0D0FF',
                                        [
                                            vm.assignment(
                                                vm.arrItemWrite('lps', vm.variable('i')),
                                                vm.variable('j'), ' // For mismatch, put match length by 1 less, than in the case of match'
                                            ),
                                            vm.assignment(
                                                undefined,
                                                vm.varPostOp('i', true)
                                            )
                                        ])
                                ]),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#F0F8FF', 'A0D0FF',
                                        [
                                            vm.assignment(
                                                vm.varWrite('j'),
                                                vm.arrItem('lps', vm.expression(vm.minus(), vm.variable('j'), vm.number(1))), ' // Note, that i is NOT incremented; only pointer j in pattern is slashed back'
                                            )
                                        ])
                                ])
                            )
                        ])
                ])
            ),
            vm.standAloneComment(),
            vm.returnStatement(vm.variable('lps'))
        ]),
        '// Preprocess the pattern (calculate lps[] array)'
    );
    const kmp = vm.functionDeclaration(
        'kmp',
        [vm.variable('txt'), vm.variable('pat')],
        vm.sequenceStatement([
            vm.assignment(
                vm.varWrite('N'),
                vm.functionCall('length', [], 'txt')
            ),
            vm.assignment(
                vm.varWrite('M'),
                vm.functionCall('length', [], 'pat')
            ),
            vm.ifStatement(
                vm.expression(vm.le(), vm.variable('M'), vm.number(0)),
                vm.returnStatement(vm.number(-1)),
                undefined
            ),
            vm.standAloneComment(),
            vm.assignment(
                vm.varWrite('lps'),
                vm.functionCall(computeLPSArray, [vm.variable('pat')])
            ),
            vm.assignment(
                vm.varWrite('j', {"role":"index", "targetArrays":["pat", "lps"]}),
                vm.number(0)
            ),
            vm.assignment(
                vm.varWrite('i', {"role":"index", "targetArrays":["txt"]}),
                vm.number(0)
            ),
            vm.standAloneComment(),
            vm.whileStatement(
                vm.expression(vm.lt(), vm.variable('i'), vm.variable('N')),
                vm.sequenceStatement([
                    vm.whileStatement(
                        vm.expression(vm.eq(), vm.arrItem('txt', vm.variable('i')), vm.arrItem('pat', vm.variable('j'))),
                        vm.sequenceStatement([
                            vm.group('// Match at position i','#F0FFF0', 'D0FFD0',
                                [
                                    vm.assignment(
                                        undefined,
                                        vm.varPostOp('j', true)
                                    ),
                                    vm.assignment(
                                        undefined,
                                        vm.varPostOp('i', true)
                                    ),
                                    vm.ifStatement(
                                        vm.expression(vm.eq(), vm.variable('j'), vm.variable('M')),
                                        vm.sequenceStatement([
                                            vm.returnStatement(vm.expression(vm.minus(), vm.variable('i'), vm.variable('j')))
                                        ]),
                                        undefined
                                    )
                                ])
                        ])
                    ),
                    vm.group('// Mismatch at position i','FFFAFA', 'FFF0F0',
                        [
                            vm.ifStatement(
                                vm.expression(vm.eq(), vm.variable('j'), vm.number(0)),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#F4F0FF', 'E0D0FF',
                                        [
                                            vm.assignment(
                                                undefined,
                                                vm.varPostOp('i', true), ' // Mismatch at the very first pattern char; cannot shift back in pattern, so start a new match'
                                            )
                                        ])
                                ]),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#F0F8FF', 'A0D0FF',
                                        [
                                            vm.assignment(
                                                vm.varWrite('j'),
                                                vm.arrItem('lps', vm.expression(vm.minus(), vm.variable('j'), vm.number(1))), ' // We know that j charaters match. Note, that i is NOT incremented.'
                                            )
                                        ])
                                ])
                            )
                        ])
                ])
            ),
            vm.standAloneComment(),
            vm.returnStatement(vm.number(-1))
        ]),
        '// Knuth-Morris-Pratt algorithm'
    );

    const usage = vm.sequenceStatement([
        vm.assignment(
            vm.varWrite('index'),
            vm.functionCall(kmp, [vm.string('ABABDABACDABABCABAB'), vm.string('ABABCABAB')])
        ),
        vm.stop()
    ]);

    return {
        code: vm.codeBlocks([computeLPSArray, kmp, usage]),
        entry: usage
    };
}();
